contracts/contracts/vault/VaultAdmin.sol

```diff
@@ -163,17 +164,156 @@ contract VaultAdmin is VaultStorage {
+        _postSwapChecks(
+            _fromAsset,
+            _fromAssetAmount,
+            fromAssetConfig,
+            _toAsset,
+            toAssetAmount,
+            _minToAssetAmount,
+            toAssetConfig,
+            config
+        );
+
+        emit Swapped(_fromAsset, _toAsset, _fromAssetAmount, toAssetAmount);
+    }
+
+    function _postSwapChecks(
+        address _fromAsset,
+        uint256 _fromAssetAmount,
+        Asset memory fromAssetConfig,
+        address _toAsset,
+        uint256 toAssetAmount,
+        uint256 _minToAssetAmount,
+        Asset memory toAssetConfig,
+        SwapConfig memory config
+    ) internal view {
+        // Check the to assets returned is above slippage amount specified by the strategist
+        require(
+            toAssetAmount >= _minToAssetAmount,
+            "Strategist slippage limit"
+        );
+
+        // Check the slippage against the Oracle in case the strategist made a mistake or has become malicious.
+        // to asset amount = from asset amount * from asset price / to asset price
+        uint256 minOracleToAssetAmount = (_fromAssetAmount *
+            (1e4 - fromAssetConfig.allowedOracleSlippageBps) *
+            IOracle(priceProvider).price(_fromAsset)) /
+            (IOracle(priceProvider).price(_toAsset) *
+                (1e4 + toAssetConfig.allowedOracleSlippageBps));
+
+        require(
+            toAssetAmount >= minOracleToAssetAmount,
+            "Oracle slippage limit exceeded"
+        );
+
+        // Check the vault's total value hasn't gone below the OToken total supply
+        // by more than the allowed percentage.
+        require(
+            (IVault(address(this)).totalValue() *
+                (1e4 + config.allowedUndervalueBps)) /
+                1e4 >=
+                oUSD.totalSupply(),
+            "Allowed value < supply"
+        );
+    }
+
+    /***************************************
+                    Swap Config
+    ****************************************/
+
+    /**
+     * @notice Set the contract the performs swaps of collateral assets.
      * @param _swapperAddr Address of the Swapper contract that implements the ISwapper interface.
      */
     function setSwapper(address _swapperAddr) external onlyGovernor {
-        swapper = _swapperAddr;
+        swapConfig.swapper = _swapperAddr;
         emit SwapperChanged(_swapperAddr);
     }
 
+    /// @notice Contract that swaps the vault's collateral assets
+    function swapper() external view returns (address swapper_) {
+        swapper_ = swapConfig.swapper;
+    }
+
+    /**
+     * @notice Set max allowed percentage the vault total value can drop below the OToken total supply in basis points
+     * when executing collateral swaps.
+     * @param _basis Percentage in basis points. eg 100 == 1%
+     */
+    function setSwapAllowedUndervalue(uint16 _basis) external onlyGovernor {
+        swapConfig.allowedUndervalueBps = _basis;
```

Comment:
We should have a require that keeps this from going over 100% / 10,000, just for the sanity of the rest of the code.

Reply:
good point. I'll add